subcommands/keys/rotate_root.go

```diff
@@ -0,0 +1,374 @@
+	Id  string
+	Key *rsa.PrivateKey
+}
+
+var doRootSync bool
+
+func init() {
+	rotate := &cobra.Command{
+		Use:     "rotate-root <offline key archive>",
+		Aliases: []string{"rotate"},
+		Short:   "Rotate root signing key used by the Factory",
+		Run:     doRotateRoot,
+		Args:    cobra.ExactArgs(1),
+	}
+	subcommands.RequireFactory(rotate)
+	rotate.Flags().BoolVarP(&doRootSync, "--sync-prod", "", false, "Make sure production root.json is up-to-date")
+	cmd.AddCommand(rotate)
+}
+
+func doRotateRoot(cmd *cobra.Command, args []string) {
+	factory := viper.GetString("factory")
+	credsFile := args[0]
+	assertWritable(credsFile)
+	creds := getOfflineCreds(credsFile)
+
+	root, err := api.TufRootGet(factory)
+	subcommands.DieNotNil(err)
+
+	subcommands.DieNotNil(syncProdRoot(factory, *root, creds))
+	if doRootSync {
+		return
+	}
+
+	curid, curPk, err := findRoot(*root, creds)
+	fmt.Println("= Current root:", curid)
+	subcommands.DieNotNil(err)
+
+	// A rotation is pretty easy:
+	// 1. change the who's listed as the root key: "swapRootKey"
+	// 2. sign the new root.json with both the old and new root
+
+	newid, newPk, newCreds := swapRootKey(root, curid, creds)
+	fmt.Println("= New root:", newid)
+
+	fmt.Println("= Resigning root.json")
+	signers := []TufSigner{
+		{Id: curid, Key: curPk},
+		{Id: newid, Key: newPk},
+	}
+	removeUnusedKeys(root)
+	subcommands.DieNotNil(signRoot(root, signers...))
+
+	bytes, err := json.MarshalIndent(root, "", "  ")
+	subcommands.DieNotNil(err)
+
+	// Create a backup before we try and commit this:
+	tmpCreds := saveTempCreds(credsFile, newCreds)
+
+	fmt.Println("= Uploading rotated root")
+	body, err := api.TufRootPost(factory, bytes)
+	if err != nil {
+		fmt.Println("\nERROR:", err)
+		fmt.Println(body)
+		fmt.Println("A temporary copy of the new root was saved:", tmpCreds)
+		fmt.Println("Before deleting this please ensure your factory isn't configured with this new key")
+		os.Exit(1)
+	}
+	if err := os.Rename(tmpCreds, credsFile); err != nil {
+		fmt.Println("\nERROR: Unable to update offline creds file.", err)
+		fmt.Println("Temp copy still available at:", tmpCreds)
+		fmt.Println("This temp file contains your new factory root private key. You must copy this file.")
+	}
+
+	// backfill this new key
+	syncProdRootOrDie(factory, *root, creds)
+}
+
+func removeUnusedKeys(root *client.TufRoot) {
+	var inuse []string
+	for _, role := range root.Signed.Roles {
+		inuse = append(inuse, role.KeyIDs...)
+	}
+	// we also have to be careful to not loose the extra root key when doing
+	// a root key rotation
+	for _, sig := range root.Signatures {
+		inuse = append(inuse, sig.KeyId)
+	}
+
+	for k := range root.Signed.Keys {
+		// is k in inuse?
+		found := false
+		for _, val := range inuse {
+			if k == val {
+				found = true
+				break
+			}
+		}
+		if !found {
+			fmt.Println("= Removing unused key:", k)
+			delete(root.Signed.Keys, k)
+		}
+	}
+}
+
+func signRoot(root *client.TufRoot, signers ...TufSigner) error {
```

Comment:
I think we may do better here from consistency point of view (what I want is upload the exact same canonical root role that we've signed).
This can be easily done with help of `tuf.Signed`:
```
func signRoot(root *client.TufRoot, signers ...TufSigner) (tuf.Signed, error) {
	meta, err := canonical.MarshalCanonical(root.Signed)
	if err != nil {
		return nil, err
	}
	signed := tuf.Signed{
		Signatures: make([]tuf.Signature, 0),
		Signed:     &json.RawMessage{},
	}
	// This line simply references a canonical JSON, no unmarshalling really happens, really fancy API.
	_ = signed.Signed.UnmarshalJSON(meta)
	// Do signing of meta here and append to signed.Signatures
	return signed, nil
}
```

Later on we might simply do this:
```
signed, err := signRoot(root, keys)
// At this line a Signed field is already pre-marshalled and will get into a resulting JSON unmodified.
// i.e. it is exactly the same as what we've signed.
body, err := json.Marshal(signed)
```

Reply:
A reason why I think it's important is we might wanna add a `--dry-run` option here similar to what I've done in `wave init` command, and I think it's really important to show a user _exactly the same root.json_ that will be delivered to devices.

Reply:
i may not be absorbing everything you are saying here, but I think what you are really wanting is to send the server the canonicaljson and not just $what-ever-json + signatures?

I used to feel the same way. It turned out to be a waste for me. Aktualizr has to canonical-json-ify what you send it to ensure security. The tuf-reposerver doesn't store the targets that way. My condensed-target stuff does this, but it turned out to just give me a false sense of security.

That said - for a future "--dryrun" option, the ability to print out canoncial-json seems sensible. But we can tack that on to this later without much trouble.

Reply:
Yeah, maybe I'm exaggregating on this.
I just tried this approach (of using Signed type) in `wave init` and it seems simple.

Will close this thread and leave it up to you.